vnext/Microsoft.ReactNative/RedBox.cpp

```diff
@@ -11,117 +11,303 @@
+          }
+        }
       }
+      showingRedBox = false;
+      showTopJSError(); // Show the next error if there is one
+    }
+  }
 
-      // Do some kind of sizing logic...
-      auto window = winrt::Windows::UI::Xaml::Window::Current();
-      auto windowBounds = window.Bounds();
-      redboxContent.MaxHeight(windowBounds.Height);
-      redboxContent.Height(windowBounds.Height);
-      redboxContent.MaxWidth(windowBounds.Width);
-      redboxContent.Width(windowBounds.Width);
-      winrt::event_token tokenSizeChanged = window.SizeChanged(
-          [redboxContent](auto const & /*sender*/, winrt::Windows::UI::Core::WindowSizeChangedEventArgs const &args) {
-            redboxContent.MaxHeight(args.Size().Height);
-            redboxContent.Height(args.Size().Height);
-            redboxContent.MaxWidth(args.Size().Width);
-            redboxContent.Width(args.Size().Width);
-          });
-
-      // On close cleanup all the events that were hooked up
-      // The closed token needs to be allocated so that it can be captured, otherwise we just capture the token before
-      // its set to anything useful
-      winrt::event_token *tokenClosed = new winrt::event_token;
-      *tokenClosed =
-          popup.Closed([ tokenDismiss, dismissButton, tokenClosed, popup, tokenSizeChanged, tokenReload, reloadButton ](
-              auto const & /*sender*/, winrt::Windows::Foundation::IInspectable const & /*args*/) noexcept {
-            dismissButton.Click(tokenDismiss);
-            reloadButton.Click(tokenReload);
-            winrt::Windows::UI::Xaml::Window::Current().SizeChanged(tokenSizeChanged);
-            popup.Closed(*tokenClosed);
-            delete tokenClosed;
-          });
-
-      std::regex colorsreg("\\x1b\\[[0-9;]*m"); // strip out console colors which is often added to JS error messages
-      errorMessageText.Text(
-          Microsoft::Common::Unicode::Utf8ToUtf16(std::regex_replace(jsExceptionInfo.exceptionMessage, colorsreg, "")));
-
-      popup.Child(redboxContent);
-      popup.IsOpen(true);
-    });
-  };
+  void showTopJSError() {
+    std::shared_ptr<RedBox> redbox;
+    {
+      std::lock_guard<std::mutex> lock(m_lockRedBox);
+      if (!m_redboxes.empty()) {
+        redbox = m_redboxes[0];
+      }
+    }
+
+    if (auto uiQueue = m_wkUIMessageQueue.lock()) {
+      if (showingRedBox || !redbox) // Only show one redbox at a time
+        return;
+      showingRedBox = true;
+      uiQueue->runOnQueue([redboxCaptured = std::move(redbox)]() { redboxCaptured->ShowNewJSError(); });
+    }
+  }
+
+  bool showingRedBox = false; // Access from UI Thread only
+  std::mutex m_lockRedBox;
+  std::vector<std::shared_ptr<RedBox>> m_redboxes; // Protected by m_lockRedBox
+  const Mso::WeakPtr<IReactHost> m_weakReactHost;
+  const std::weak_ptr<facebook::react::MessageQueueThread> m_wkUIMessageQueue;
+};
+
+std::shared_ptr<IRedBoxHandler> CreateRedBoxHandler(
+    Mso::WeakPtr<IReactHost> weakReactHost,
```

Comment:
>weakReactHost [](start = 29, length = 13)

In Office we have an OACR rule that says that values of types with non-trivial destructor must not be passed by value. The reason is that it generates quite inefficient code where the method must have the destructor call for such parameters even if they are just 'pass-only' as here.
Please consider changing the parameter types either to 'const&' or '&&'. The '&&' is good to communicate that the function or constructor is going to hold or own the passed value. While 'const&' is that it only temporary 'looks' at it.